include/circt/Dialect/FIRRTL/InstanceGraph.h

```diff
@@ -0,0 +1,148 @@
+//===- InstanceGraph.h - Instance graph -------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the FIRRTL InstanceGraph, which is similar to a CallGraph.
+//
+//===----------------------------------------------------------------------===//
+#ifndef CIRCT_DIALECT_FIRRTL_INSTANCEGRAPH_H
+#define CIRCT_DIALECT_FIRRTL_INSTANCEGRAPH_H
+
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/OperationSupport.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/SmallVector.h"
+#include <memory>
+
+namespace circt {
+namespace firrtl {
+
+class InstanceGraphNode {
+  using EdgeSetT = llvm::SmallVector<InstanceGraphNode *, 1>;
+
+public:
+  InstanceGraphNode(Operation *module) : module(module) {}
+  InstanceGraphNode(InstanceGraphNode &&other)
+      : module(other.module), instances(std::move(other.instances)) {
+    other.module = nullptr;
+  }
+
+  /// Get the module that this node is tracking.
+  Operation *getModule() const { return module; }
+
+  using iterator = EdgeSetT::iterator;
+  iterator begin() { return instances.begin(); }
+  iterator end() { return instances.end(); }
+
+  using const_iterator = EdgeSetT::const_iterator;
+  const_iterator begin() const { return instances.begin(); }
+  const_iterator end() const { return instances.end(); }
```

Comment:
Just a suggestion, but I'd dump all const'ness support here, because it is meaningless because C++ doesn't have a mutation model that works for graphs (a const iterator still products pointers to mutable nodes).  The right thing is something like:

`using iterator = EdgeSetT::const_iterator;`

This prevents clients from mutating the edge list with the iterator (not something anyone should want) but still reflects that the dereferenced nodes are mutable.

Reply:
xref: https://mlir.llvm.org/docs/Rationale/UsageOfConst/

Reply:
I dropped all the weird constness stuff. I think I did it correctly with `using iterator = EdgeVec::const_iterator`, but I should go over it once more.